backend/src/prereq_loader.ts

```diff
@@ -0,0 +1,250 @@
+  yearObj: ScheduleYear,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleYear> {
+  return {
+    year: yearObj.year,
+    fall: await prereqifyScheduleTerm(yearObj.fall, loader),
+    spring: await prereqifyScheduleTerm(yearObj.spring, loader),
+    summer1: await prereqifyScheduleTerm(yearObj.summer1, loader),
+    summer2: await prereqifyScheduleTerm(yearObj.summer2, loader),
+    isSummerFull: yearObj.isSummerFull,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a term's courses.
+ * @param termObj the term object to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleTerm(
+  termObj: ScheduleTerm,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleTerm> {
+  // the new classes.
+  const newClasses: Promise<ScheduleCourse[]> = Promise.all(
+    termObj.classes.map((course: ScheduleCourse) =>
+      prereqifyScheduleCourse(course, termObj.termId, loader)
+    )
+  );
+
+  return {
+    season: termObj.season,
+    year: termObj.year,
+    termId: termObj.termId,
+    id: termObj.id,
+    classes: await newClasses,
+    status: termObj.status,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a provided course.
+ * @param course the course to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleCourse(
+  course: ScheduleCourse,
+  termId: number,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleCourse> {
+  // the base prereqified object
+  const prereqified: ScheduleCourse = {
+    classId: course.classId,
+    subject: course.subject,
+    numCreditsMin: course.numCreditsMin,
+    numCreditsMax: course.numCreditsMax,
+  };
+
+  let queryResult: PrereqQueryResult;
+  try {
+    // produces the class.
+    queryResult = await loader.load({
+      subject: course.subject,
+      classId: Number(course.classId),
+      termId: termId,
+    });
+  } catch (err) {
+    throw err;
+    // if we error, then return the previous course.
+    return course;
+  }
+
+  // optionally add prereqs, coreqs to object.
+  if (queryResult) {
+    queryResult.coreqs ? (prereqified.coreqs = queryResult.coreqs) : undefined;
+    queryResult.prereqs
+      ? (prereqified.prereqs = queryResult.prereqs)
+      : undefined;
+    // prereqified.numCreditsMax = queryResult.maxCredits;
+    // prereqified.numCreditsMin = queryResult.minCredits;
+    return prereqified;
+  } else {
+    return course;
+  }
+}
+
+/**
+ * Queries SearchNEU using GraphQL to look up the prereqs for each of the provided courses.
+ * @param courses the courses to lookup prereqs for
+ */
+async function queryCoursePrereqData(
+  courses: SimpleCourse[]
+): Promise<PrereqQueryResult[]> {
+  // for each one of the courses, map to a string.
+  const courseSchema: string[] = courses.map((course: SimpleCourse) => {
+    return `class(classId: ${course.classId}, subject: "${course.subject}") { 
+      occurrence(termId: ${202010}) {
```

Comment:
why is the termId set to `202010` here specifically?

Reply:
Given a plan with courses that don't have prereqs, I decided that we should look at the most recent year of the catalog to grab their prereqs.

However, I'm going to change this to be year-specific (pass in a year along with the plan to grab prereqs for that plan, from that year of the catalog).